drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+			data->async->rx_cnt.ppi,
+			nrf_uarte_event_address_get(cfg->uarte_regs,
+						    NRF_UARTE_EVENT_RXDRDY),
+			nrfx_timer_task_address_get(&data->async->timer,
+						    NRF_TIMER_TASK_COUNT));
+
+		if (ret != NRFX_SUCCESS) {
+			return -EIO;
+		}
+#else
+		nrf_uarte_publish_set(cfg->uarte_regs,
+				      NRF_UARTE_EVENT_RXDRDY,
+				      data->async->rx_cnt.ppi);
+		nrf_timer_subscribe_set(data->async->timer.p_reg,
+					NRF_TIMER_TASK_COUNT,
+					data->async->rx_cnt.ppi);
+#endif
+	} else {
+		nrf_uarte_int_enable(cfg->uarte_regs,
+				     NRF_UARTE_INT_RXDRDY_MASK);
+	}
+	data->async->rx_total_byte_cnt = 0;
+
+	nrf_uarte_int_enable(cfg->uarte_regs,
+			     NRF_UARTE_INT_ENDRX_MASK |
+			     NRF_UARTE_INT_RXSTARTED_MASK |
+			     NRF_UARTE_INT_ERROR_MASK |
+			     NRF_UARTE_INT_ENDTX_MASK |
+			     NRF_UARTE_INT_TXSTOPPED_MASK |
+			     NRF_UARTE_INT_RXTO_MASK);
+	nrf_uarte_enable(cfg->uarte_regs);
+
+	k_delayed_work_init(&data->async->rx_timeout_work,
+			    data->async->rx_timeout_handler);
+	k_delayed_work_init(&data->async->tx_timeout_work,
+			    data->async->tx_timeout_handler);
+
+	return 0;
+}
+
+static int uarte_nrfx_rx_buf_rsp(struct device *dev, u8_t *buf, size_t len)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (data->async->rx_next_buf == NULL) {
+		data->async->rx_next_buf = buf;
+		nrf_uarte_rx_buffer_set(cfg->uarte_regs, buf, len);
+	} else {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+
+static int uarte_nrfx_rx_enable(struct device *dev, u8_t *buf, size_t len,
+				u32_t timeout)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	nrfx_err_t err;
+
+	if (hw_rx_counting_enabled(data)) {
+		err = gppi_channel_enable(data->async->rx_cnt.ppi);
+		if (err != NRFX_SUCCESS) {
+			return -EIO;
+		}
+	}
+
+	data->async->rx_timeout = timeout;
+	data->async->rx_timeout_slab = timeout / RX_TIMEOUT_DIV > MS_PER_TICK ?
+				       (timeout / RX_TIMEOUT_DIV) - 1 :
+				       MS_PER_TICK;
+	data->async->rx_timeout_slab_real = data->async->rx_timeout_slab
+				      + MS_PER_TICK;
+	data->async->rx_size = len;
+	data->async->rx_buf = buf;
+	nrf_uarte_rx_buffer_set(cfg->uarte_regs, buf, len);
+
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXSTARTED);
+	if (!hw_rx_counting_enabled(data)) {
+		nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXDRDY);
+	}
+
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTRX);
+	data->async->rx_enabled = true;
+	return 0;
+}
+
+#define UARTE_ERROR_FROM_MASK(mask) \
+	(mask & NRF_UARTE_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : mask & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : mask & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : mask & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+static void error_isr(struct device *dev)
+{
```

Comment:
The reception should be stopped in this handler. It is not done automatically by hardware when the `ERROR` event occurs.